File: c++/_560_SubArraySumEqualsK.cc
//
// Created by ljt on 2026/5/27.
//
#include <iostream>
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
    /**
     * 解题思路：假设有两个位置i和j，prefix_sum为前缀和的数组，其中j > i，那么子串的subarr[i,j]的和为prefix_sum[j] - prefix_sum[i - 1], 那么如果要求出子数组和等于k的子数组数量，这道问题就变成了如下的形式
     * k = prefix_sum[j] - prefix_sum[i - 1] => prefix_sum[j] = k + prefix_sum[i - 1], 那么只需要prefix_sum[i - 1] = prefix_sum[j] - k，这个式子就可以成立，这个就很清晰了，
     * 我们维护一个map，该map的key value分别为 prefix_sum -> count，当我们遍历到了j的位置的时候，只需要检查一下prefix_sum[j] - k的和的个数，相当于就算出了子数组和等于k的个数
     * 该题巧妙的将每次根据前缀和计算子数组和的O(n^2)算法转换成了O(1)的算法
     * @param nums
     * @param k
     * @return
     */
    int subarraySum(vector<int> &nums, int k) {
        if (nums.empty()) {
            return 0;
        }

        std::map<int, int> m;
        m[0] = 1; // 这里的初始化是为了处理边界情况，即[0,j]的前缀刚好等于k
        int sum = 0;
        int result = 0;
        for (auto &val: nums) {
            sum += val;
            result += m[sum - k];
            m[sum]++;
        }
        return result;
    }
};

int main() {
    Solution s;
    std::vector<int> nums{1, 1, 1};
    int k = 2;
    auto result = s.subarraySum(nums, k); // 2
    printf("%d\n", result);

    nums = {1, 2, 3};
    k = 3;
    result = s.subarraySum(nums, k); // 2
    printf("%d\n", result);

    nums = {1, 2, 3, 4, -4, 0};
    k = 3;
    result = s.subarraySum(nums, k); // 4
    printf("%d\n", result);
    return 0;
}

File: go/560_SubArraySumEqualsK.go
package main

import "fmt"

/*
*
解题思路：返回和为k的子数组数量
解法1(容易想出来的暴力解法)
维护一个同样长度的数组，该数组记录了0-i之间的和，记为sum，这样子任意区间和的计算公式如下：

	k = sum[j] - sum[i - 1]

该方法过慢，解法2是对其的一个加速
解法2（优化解法1）
依旧是前缀和的思路，将解法的式子进行一个移项，得到
sum[i-1] = sum[j] - k
上面这个等式意味着，任意的前缀和只要能找到其前面前缀和-k的个数，我们就相当于找到了满足和为k的子数组的数量，而这个只需要遍历一遍就能得到结果，代价就是需要一个hash表来维护前缀和等于x的个数
*/
func main() {
	var (
		nums = []int{1, 1, 1}
		k    = 2
	)
	result := subarraySum(nums, k)
	result2 := subarraySum2(nums, k)
	fmt.Println(result, "--------", result2) // 2

	nums = []int{1, 2, 3}
	k = 3
	result = subarraySum(nums, k)
	result2 = subarraySum2(nums, k)
	fmt.Println(result, "--------", result2) // 2

	nums = []int{1, 2, 3, 4, -4, 0}
	k = 3
	result = subarraySum(nums, k)
	result2 = subarraySum2(nums, k)
	fmt.Println(result, "--------", result2) // 4
}

func subarraySum(nums []int, k int) int {
	if len(nums) == 0 {
		return 0
	}
	var (
		sum    = make([]int, len(nums))
		tmp    = 0
		result = 0
	)
	for idx, num := range nums {
		tmp += num
		sum[idx] = tmp
		if tmp == k {
			result++
		}
	}

	// 有了前缀和之后，开始计算任意区间的子串和
	for i := 1; i < len(nums); i++ {
		if nums[i] == k {
			result++
		}
		// 开始计算子串的和
		for j := i + 1; j < len(nums); j++ {
			tmp = sum[j] - sum[i-1]
			if tmp == k {
				result++
			}
		}
	}
	return result
}

// subarraySum2 对解法1的优化，使用一个hash表记录前缀和的个数
func subarraySum2(nums []int, k int) int {
	if len(nums) == 0 {
		return 0
	}
	var (
		m      = make(map[int]int, 0)
		prefix = 0
		result = 0
	)
	m[0] = 1
	for _, num := range nums {
		prefix += num
		result += m[prefix-k]
		m[prefix]++
	}
	return result
}
